Configure dagster env vars from settings if not set already

zschira opened the pull request:
This PR addresses #2301 by setting `PUDL_OUTPUT` and `PUDL_CACHE` from `pudl_settings` if these env vars are not already set.

zschira commented on src/pudl/workspace/setup.py:
```diff
@@ -67,7 +67,9 @@ def get_defaults():
     # specified absolute paths out when we read it:
     pudl_in = pathlib.Path(default_workspace["pudl_in"]).expanduser().resolve()
-    pudl_out = pathlib.Path(default_workspace["pudl_out"]).expanduser().resolve()
+    pudl_out = (
```
This combines the output directories instead of having `sqlite` and `parquet` dirs. That does mean people with an existing workspace will need to add an `output` directory at the top level of their workspace.

jdangerx replied:
So we'll have `workspace/output/sqlite` and `workspace/output/parquet`?

zschira replied:
No we'll just have a single `workspace/output` directory, and the `sqlite_dir` and `parquet_dir` params will point there

bendnorman reviewed (changes requested):
Looking good! Dagit throws errors from the `EnvVar` when `PUDL_OUTPUT` and `PUDL_CACHE` aren't set. I think we need to replace the config schemas that use the `EnvVar` class with a helper function that tries to pull the output path from the env vars and falls back to the pudl_settings default if the var isn't set. Something like this:

```python
def get_pudl_output_default():
    return os.environ.get("PUDL_OUTPUT", pudl.workspace.setup.get_defaults()["pudl_out"])


@io_manager(
    config_schema={
        "pudl_output_path": Field(
            str,
            description="Path of directory to store the database in.",
            default_value=get_pudl_output_default(),
        ),
    }
)
```
What do you think @zschira?

bendnorman commented:
Also, should we add deprecation warnings to the functions in `pudl.workspace.setup`?

I think the desired end state is to remove all of `pudl.workspace.setup` and force users to set `PUDL_OUTPUT` and `PUDL_CACHE`.

zaneselvans commented:
I imagine you're already thinking about this or already ran into it and fixed it but I just wanted to flag it in case: there are/were some circumstances in which we change the contents of `pudl_settings` to affect where the inputs are read from, or outputs are written to, and if the `PUDL_CACHE` and `PUDL_OUTPUT` envvars are getting set globally and not updated along with those edits to the `pudl_settings` it seems like there could be some unexpected outcomes.

Mostly I think this is happening in the context of running the tests, e.g.:
* When telling the tests to download all the data itself into a temporary directory, rather than use an existing datastore.
* When telling the tests to use a pre-existing set of live DBs rather than running the ETL to generate its own DB for testing purposes in a temporary directory.
* When telling the tests to put outputs in a non-standard location to avoid clobbering someone's existing live DB.
* When telling the tests to put outputs in a non-standard location to avoid clobbering DBs generated by other tests.

zschira commented on src/pudl/helpers.py:
```diff
@@ -1589,7 +1590,10 @@ def post_process(self, value: str) -> str:
         if value is None:
             try:
-                return os.environ[self.env_var]
+                value = os.environ.get(self.env_var)
```
using `os.environ.get(self.env_var, get_defaults()[self.env_var])` will evaluate `get_defaults`, which causes the CI to fail when no `.pudl.yaml` file exists, so I've separated these.

jdangerx replied:
Easier to read, too!

zschira commented:
@zaneselvans we added fixtures to set `PUDL_CACHE` and `PUDL_OUTPUT` in a way that mirrors how the `pudl_settings` were being handled before, so I think that should all be accounted for.

jdangerx reviewed (approved):
Sweet! I feel... OK about ignoring those coverage issues for now. But only OK.

bendnorman commented on src/pudl/cli.py:
```diff
@@ -86,14 +87,23 @@ def main():
         logfile=args.logfile, loglevel=args.loglevel
     )
 
+    etl_settings = EtlSettings.from_yaml(args.settings_file)
+
+    if (not os.getenv("PUDL_OUT")) or (not os.getenv("PUDL_CACHE")):
+        pudl_settings = pudl.workspace.setup.derive_paths(
+            pudl_in=etl_settings.pudl_in, pudl_out=etl_settings.pudl_out
+        )
+
+        os.environ["PUDL_CACHE"] = pudl_settings["data_dir"]
+        os.environ["PUDL_OUTPUT"] = pudl_settings["pudl_out"]
+        os.environ["DAGSTER_HOME"] = pudl_settings["pudl_in"]
+
```
We might want to add a comment or a deprecation warning here that explains why we're falling back to `pudl_settings` if the env vars aren't set.

bendnorman commented on src/pudl/workspace/setup.py:
```diff
@@ -110,7 +112,11 @@ def derive_paths(pudl_in, pudl_out):
     # One directory per output format:
     for fmt in ["sqlite", "parquet"]:
-        pudl_settings[f"{fmt}_dir"] = str(pudl_out / fmt)
+        pudl_settings[f"{fmt}_dir"] = str(pudl_out)
```
Should we add a deprecation warning here about the sqlite and parquet directories?

bendnorman reviewed (changes requested):
Looks good! I like your solution for the EnvVar problem. I think this PR just needs some deprecation warning and it'll be good to go.

zschira commented:
@bendnorman I added some deprecation warnings. I pointed them all to the [dev setup](https://catalystcoop-pudl.readthedocs.io/en/dev/dev/dev_setup.html) portion of the docs because I figure that's where we'll mention the new env vars. I'm probably going to run for the day, but feel free to reword/edit these if you'd like.

@jdangerx I don't feel great about the test coverage, but also all the code I added can be deleted when we deprecate the `pudl_settings`, so it doesn't feel too high priority to me.

bendnorman commented on src/pudl/helpers.py:
```diff
@@ -1589,7 +1590,10 @@ def post_process(self, value: str) -> str:
+
+                if not value:
+                    value = get_defaults()[self.env_var]
             except KeyError:
```
I guess this KeyError isn't relevant anymore? Should we keep it around for when we deprecate pudl_settings?
